server/src/main/java/io/crate/metadata/TransactionContext.java

```diff
@@ -23,38 +23,77 @@
 package io.crate.metadata;
 
 import io.crate.metadata.settings.SessionSettings;
-import org.joda.time.DateTimeUtils;
 
-public interface TransactionContext {
+import java.time.Clock;
+import java.time.Instant;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
 
-    static TransactionContext of(SessionSettings sessionSettings) {
-        return new StaticTransactionContext(sessionSettings);
+public class TransactionContext {
+
+    public static TransactionContext of(SessionSettings sessionSettings) {
+        return new TransactionContext(sessionSettings);
     }
 
-    long currentTimeMillis();
+    private static final AtomicReference<Supplier<Long>> MILLIS_PROVIDER = new AtomicReference<>(null);
```

Comment:
I think it would be cool if we could de-couple more of the time related logic into more specific components. The TransactionContext should only be concerned about caching the value retrieved to ensure the same value is re-used across the transaction. It shouldn't know too much about how that value is produced.

For example we could introduce a `Clocks` class that encapsulates the logic about fixing/mocking the current time.
Maybe even work with the `Clock` classes directly instead of the `Supplier<Long>`.

There is both

```
Clock.systemUTC();
Clock.fixed(..);
```